src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,293 @@
+/***********************************************************************
+ * Distributed under the MIT software license, see the accompanying    *
+ * file COPYING or https://www.opensource.org/licenses/mit-license.php.*
+ ***********************************************************************/
+
+#ifndef SECP256K1_MODULE_MUSIG_KEYAGG_IMPL_H
+#define SECP256K1_MODULE_MUSIG_KEYAGG_IMPL_H
+
+#include <string.h>
+
+#include "keyagg.h"
+#include "../../eckey.h"
+#include "../../ecmult.h"
+#include "../../field.h"
+#include "../../group.h"
+#include "../../hash.h"
+#include "../../util.h"
+
+static const unsigned char secp256k1_musig_keyagg_cache_magic[4] = { 0xf4, 0xad, 0xbb, 0xdf };
+
+/* A keyagg cache consists of
+ * - 4 byte magic set during initialization to allow detecting an uninitialized
+ *   object.
+ * - 64 byte aggregate (and potentially tweaked) public key
+ * - 64 byte "second" public key (set to the point at infinity if not present)
+ * - 32 byte hash of all public keys
+ * - 1 byte the parity of the internal key (if tweaked, otherwise 0)
+ * - 32 byte tweak
+ */
+/* Requires that cache_i->pk is not infinity. */
+static void secp256k1_keyagg_cache_save(secp256k1_musig_keyagg_cache *cache, secp256k1_keyagg_cache_internal *cache_i) {
+    unsigned char *ptr = cache->data;
+    memcpy(ptr, secp256k1_musig_keyagg_cache_magic, 4);
+    ptr += 4;
+    secp256k1_ge_to_bytes(ptr, &cache_i->pk);
+    ptr += 64;
+    secp256k1_ge_to_bytes_ext(ptr, &cache_i->second_pk);
+    ptr += 64;
+    memcpy(ptr, cache_i->pk_hash, 32);
+    ptr += 32;
+    *ptr = cache_i->parity_acc;
+    ptr += 1;
+    secp256k1_scalar_get_b32(ptr, &cache_i->tweak);
+}
+
+static int secp256k1_keyagg_cache_load(const secp256k1_context* ctx, secp256k1_keyagg_cache_internal *cache_i, const secp256k1_musig_keyagg_cache *cache) {
```

Comment:
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Note to self/reviewers: wasn't obvious to me how ctx was being used at first but it's being used in the `ARG_CHECK` macro.

Reply:
On second thought, seems a bit strange to be using `ARG_CHECK` inside an internal function (i.e., not exposed in the musig module API). Why not do an assert or a return here, which would let you simplify the function by removing `ctx` from the function signature?